### PR TITLE
Allow kDOPs as the user input through AccessTraits

### LINKED ISSUE #867: Boundary volume hierarchy for kDOPs
We are currently using ArborX to perform a coarse collision search in our finite element code. Specifically, we create kDOPs which approximate the convex hull of our possible collision partners and then perform an intersection search with ArborX. The pairs obtained are then checked for actual intersection, which is an expensive operation, therefore we want to keep the number of pairs found by ArborX as small as possible. For the type of bodies we consider, we get good results by using the volume type `ArborX::Experimental::KDOP<26>`. According to the [documentation](https://github.com/arborx/ArborX/wiki/ArborX%3A%3ABoundingVolumeHierarchy%3A%3ABoundingVolumeHierarchy), the primitive type for the BVH has to decay to either `AborX::Point` or `ArborX::Box`. Therefore, we convert the primitive kDOPs to boxes in the creation of the BVH.

This leads to the following issue:
kdop
In this example we expect an intersection between primitive 0 and predicate 0. However, since we convert all primitives to boxes, we also get an intersection primitive 1 and predicate 0. We solve this by running a post-processing step where all found _box-kDOP_ pairs are checked for the intended _kDOP-kDOP_ intersection.

Is it possible to include the kDOP information in the creation of the BVH, so that `query` already searches against the kDOPs of the primitives? This would make our post-processing obsolete, and simplify our implementation, especially for distributed trees. There is a base version of BVH which is templated on the bounding volume type `template <typename MemorySpace, typename BoundingVolume = Box, typename Enable = void> class BasicBoundingVolumeHierarchy` but we could not get that one to work.

I attached a minimal working example of our code (works with ArborX v1.4), which corresponds to the figure. (Also feel free to comment on possible improvements to our ArborX implementation.)

Thanks!

```c++
#include <ArborX_KDOP.hpp>
#include <ArborX_LinearBVH.hpp>

using ExecutionSpace = Kokkos::DefaultExecutionSpace;
using MemorySpace = ExecutionSpace::memory_space;
using kdop = ArborX::Experimental::KDOP<26>;

template <>
struct ArborX::AccessTraits<std::vector<kdop>, ArborX::PrimitivesTag>
{
  using memory_space = MemorySpace;
  using size_type = typename MemorySpace::size_type;
  static KOKKOS_FUNCTION size_type size(std::vector<kdop> const &volumes)
  {
    return volumes.size();
  }
  static KOKKOS_FUNCTION auto get(std::vector<kdop> const &volumes, size_type i)
  {
    return ArborX::Box{volumes[i]};
  }
};

template <>
struct ArborX::AccessTraits<std::vector<kdop>, ArborX::PredicatesTag>
{
  using memory_space = MemorySpace;
  using size_type = typename MemorySpace::size_type;
  static KOKKOS_FUNCTION size_type size(std::vector<kdop> const &volumes)
  {
    return volumes.size();
  }
  static KOKKOS_FUNCTION auto get(std::vector<kdop> const &volumes, size_type i)
  {
    return attach(intersects(volumes[i]), i);
  }
};

template <typename T>
void print_pairs(const T &indices, const T &offset)
{

  for (size_t i_offset = 0; i_offset < offset.size() - 1; i_offset++)
  {
    for (int j = offset[i_offset]; j < offset[i_offset + 1]; j++)
    {
      std::cout << "\n  predicate " << i_offset << " intersects with primitive "
                << indices[j];
    }
  }
}

int main(int argc, char *argv[])
{
  Kokkos::ScopeGuard guard(argc, argv);

  // Create two vectors with bounding volumes
  std::vector<kdop> volumes_1;
  std::vector<kdop> volumes_2;
  {
    kdop tmp;
    tmp += ArborX::Point{0.0, 0.0, 0.0};
    tmp += ArborX::Point{1.0, 1.0, 0.0};
    tmp += ArborX::Point{0.5, 0.0, 0.0};
    volumes_1.push_back(tmp);
  }
  {
    kdop tmp;
    tmp += ArborX::Point{1.0, -1.0, 0.0};
    tmp += ArborX::Point{1.0, 0.25, 0.0};
    tmp += ArborX::Point{0.75, 0.0, 0.0};
    volumes_1.push_back(tmp);
  }
  {
    kdop tmp;
    tmp += ArborX::Point{0.5, 0.25, 0.0};
    tmp += ArborX::Point{0.8, 0.25, 0.0};
    tmp += ArborX::Point{0.75, 0.125, 0.0};
    volumes_2.push_back(tmp);
  }

  // Construct BVH and perform intersection search
  ExecutionSpace space{};
  ArborX::BoundingVolumeHierarchy<MemorySpace> bvh{space, volumes_1};
  Kokkos::View<int *, ExecutionSpace> indices("indices", 0);
  Kokkos::View<int *, ExecutionSpace> offset("offset", 0);
  bvh.query(space, volumes_2, indices, offset);

  // Print the results from the query
  std::cout << "bvh.query (BOX vs kDOP) yields the following intersections:";
  print_pairs(indices, offset);

  // Perform a post-processing to check kDOP vs kDOP
  std::vector<int> indices_final;
  std::vector<int> offsets_final;
  offsets_final.push_back(0);
  for (size_t i_predicate = 0; i_predicate < volumes_2.size(); i_predicate++)
  {
    int collisions_for_predicate_i = 0;
    auto const &predicate = volumes_2[i_predicate];
    for (int j = offset[i_predicate]; j < offset[i_predicate + 1]; j++)
    {
      int i_primitive = indices[j];
      if (ArborX::Experimental::intersects(predicate, volumes_1[i_primitive]))
      {
        indices_final.push_back(i_primitive);
        collisions_for_predicate_i += 1;
      }
    }
    offsets_final.push_back(offsets_final.back() + collisions_for_predicate_i);
  }

  // Print the post-processed results
  std::cout << "\nkDOP vs kDOP yields the following intersections:";
  print_pairs(indices_final, offsets_final);
  std::cout << "\n";

  return 0;
}

// Expected output:
// > bvh.query (BOX vs kDOP) yields the following intersections:
// >   predicate 0 intersects with primitive 1
// >   predicate 0 intersects with primitive 0
// > kDOP vs kDOP yields the following intersections:
// >   predicate 0 intersects with primitive 0
```

### PR DESCRIPTION
**Pull request by aprokop**
Fix #867.

### CODE REVIEW

File: src/details/ArborX_AccessTraits.hpp [hunk 1]
```diff
@@ -187,9 +187,10 @@ void check_valid_access_traits(PrimitivesTag, Primitives const &)
       "member function");
   using T = std::decay_t<Kokkos::detected_t<AccessTraitsGetArchetypeExpression,
                                             Access, Primitives>>;
-  static_assert(GeometryTraits::is_point<T>{} || GeometryTraits::is_box<T>{},
+  static_assert(GeometryTraits::is_point<T>{} || GeometryTraits::is_box<T>{} ||
+                    GeometryTraits::is_kdop<T>{},
                 "AccessTraits<Primitives,PrimitivesTag>::get() return type "
-                "must decay to a point or a box type");
+                "must decay to a point, a box or a kdop type");
```
Comment (masterleinad) [inline src/details/ArborX_AccessTraits.hpp:193]:
```suggestion
                "must decay to a point, a box or a k-DOP type");
```

**Review by masterleinad — COMMENTED** [review]
We should add a test for this.

**Comment by aprokop** [comment]
This seems to do what we want to. I'm not 100% certain if I did it correctly, though. Mainly, the switch to using `expand` in scene bounding box calculation. If that's done, then it does necessitate the move of some calculations in the algorithms.

**Comment by aprokop** [comment]
This PR bypasses the question whether we allow users to easily use a hierarchy based on k-DOPs, and simply allows us to do that internally.

**Pull request closed without merging**